# Task 2: Recommendation Engine - Skeleton Notebook

This notebook provides a very basic example for the notebook you are expected to submit for Task 2 of the Final Project. The main purpose is that we can try different examples to get a better sense of your approach. Compared to Task 1 (Kaggle Competition), we don't have any objective means to evaluate the recommendations. 

Some general comments:
* You can import any data you need. This particularly includes your cleaned version of the properties dataset (incl. the auxiliary data or any other data you might have collected); there's no need to show the data cleaning / preprocessing steps in this notebook.
* You can also import your code in form of external Python (.py) script. You're actually encouraged to do so to keep this notebook light and uncluttered.
* **Important:** Please consider this notebook as an example and not to set specific requirements. Your notebook is likely to look very different. As long there is a section where we can easily test your solution, it should be fine.

### Group 3: Pretty Soldier Sailor Moon
Members: SHI Jingyi, MIAO Di, WU Zhanxin, ZHENG Wenyue

## Setting up the Notebook

In [14]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.metrics.pairwise import cosine_similarity

from sklearn import preprocessing

#import turicreate as tc
#turicreate only for track 3

## Load the Data

For this example, we use a simplified version of the dataset with only 2k+ data samples, each with only a subset of features.

In [16]:
df_sample = pd.read_csv('data/preprocessed_train.csv')

df_sample.head()

listing_id                                              title  \
0      259374  hdb flat for sale in 506b serangoon north aven...   
1      665422                 4 bed condo for sale in meyerhouse   
2      857699               3 bed condo for sale in leedon green   
3      216061                 2 bed condo for sale in one bernam   
4      625939       2 bed condo for sale in forett @ bukit timah   

                              address         property_name property_type  \
0  hougang / punggol / sengkang (d19)  hdb-serangoon estate           hdb   
1                      128 meyer road            meyerhouse         condo   
2                   26 leedon heights          leedon green         condo   
3                     1 bernam street            one bernam         condo   
4                    32 toh tuck road  forett @ bukit timah         condo   

              tenure  built_year  num_beds  num_baths  size_sqft  ...  \
0  99-year leasehold      1992.0         4          2       1575  ...   
1           freehold      2022.0         4          6       3070  ...   
2           freehold      2023.0         3          2        958  ...   
3  99-year leasehold      2026.0         2          1        732  ...   
4           freehold      2024.0         2          2        689  ...   

  elevation            subzone  planning_area      price  bed_bath_ratio  \
0         0    serangoon north      serangoon   995400.0        2.000000   
1         0        mountbatten  marine parade  8485000.0        0.666667   
2         0       farrer court    bukit timah  2626000.0        1.500000   
3         0              anson  downtown core  1764000.0        2.000000   
4         0  bukit batok south    bukit batok  1702000.0        1.000000   

   count_commerical  count_mrt_station count_primary count_secondary  \
0                 0                  0             3               4   
1                 0                  0             1               1   
2                 0                  2             0               0   
3                 0                  3             1               0   
4                 0                  0             1               0   

   count_shopping  
0               1  
1               1  
2               1  
3               1  
4               0  

[5 rows x 26 columns]

---
#Compute Similarity


In [17]:
df_recom = df_sample.copy()
df_recom = df_recom.drop(columns = ['title','address','property_name','property_details_url','available_unit_types','subzone'])
df_recom.sample(5)

listing_id property_type             tenure  built_year  num_beds  \
2239       857663         condo           freehold      2022.0         4   
1179       244926   hdb 4 rooms  99-year leasehold      1983.0         3   
10782      244411         condo  99-year leasehold      2026.0         2   
7674       161259         condo  99-year leasehold      2024.0         2   
13848      405041   hdb 3 rooms  99-year leasehold      2006.0         3   

       num_baths  size_sqft   furnishing  total_num_units       lat  \
2239           4       1711  unspecified       250.000000  1.344334   
1179           2       1011      partial       702.066667  1.346136   
10782          2        818  unspecified       407.000000  1.310640   
7674           2        753  unspecified       298.000000  1.299881   
13848          2       1184  unspecified       610.306122  1.383239   

              lng  elevation  planning_area      price  bed_bath_ratio  \
2239   103.878690          0      serangoon  3549000.0             1.0   
1179   103.734150          0    jurong east   462000.0             1.5   
10782  103.852149          0        kallang  1822800.0             1.0   
7674   103.888114          0  marine parade  2277100.0             1.0   
13848  103.897653          0       sengkang   629000.0             1.5   

       count_commerical  count_mrt_station  count_primary  count_secondary  \
2239                  0                  3              3                5   
1179                  0                  1              2                3   
10782                 0                  5              2                0   
7674                  0                  2              0                3   
13848                 0                  2              4                2   

       count_shopping  
2239                1  
1179                0  
10782               4  
7674                0  
13848               1

In [18]:
df_dummy = pd.get_dummies(df_recom.drop(columns = 'listing_id'))
df_dummy = preprocessing.scale(df_dummy)

In [19]:
house_similarities = cosine_similarity(df_dummy, df_dummy)
pd.DataFrame(house_similarities).head()

0         1         2         3         4         5         6      \
0  1.000000 -0.217255 -0.233677 -0.000384 -0.173221 -0.063766 -0.132751   
1 -0.217255  1.000000  0.238268 -0.190450 -0.007474 -0.144756  0.134546   
2 -0.233677  0.238268  1.000000  0.036283  0.162182 -0.050507  0.138891   
3 -0.000384 -0.190450  0.036283  1.000000  0.023194  0.193900 -0.157820   
4 -0.173221 -0.007474  0.162182  0.023194  1.000000 -0.025068 -0.052404   

      7         8         9      ...     18318     18319     18320     18321  \
0 -0.113384 -0.053279  0.201103  ...  0.001060  0.006448 -0.170127 -0.259176   
1  0.013270 -0.109619 -0.150231  ... -0.016162 -0.135100 -0.039292 -0.047918   
2  0.108607  0.720528 -0.143064  ...  0.046366  0.034607  0.044750  0.715588   
3  0.064167  0.224893  0.172345  ... -0.118946  0.106918  0.041412  0.108371   
4  0.208573  0.211388 -0.127641  ... -0.063063 -0.029147  0.156218  0.320328   

      18322     18323     18324     18325     18326     18327  
0 -0.099561 -0.082712 -0.149036 -0.200056  0.069160 -0.240470  
1  0.163807 -0.145048  0.058021  0.159760 -0.152115  0.364600  
2 -0.049380 -0.032911  0.063152  0.024847 -0.105892  0.309795  
3 -0.138608  0.079971  0.025560 -0.104755 -0.090339 -0.159480  
4 -0.029493  0.111331  0.151559  0.087338 -0.087044  0.016971  

[5 rows x 18328 columns]

In [20]:
#example for the n th record
n=1234
one_sim = house_similarities[n]
one_sim = np.delete(one_sim,obj = n)
k = 5
one_sim[one_sim.argsort()[-k:len(one_sim)]]

array([0.99999874, 0.99999874, 0.99999874, 0.99999874, 0.99999901])

# Track1 
### Pairwise Item Similarities
## Computing the Top Recommendations

The method `get_top_recommendations()` shows an example of how to get the top recommendations for a given data sample (data sample = row in the dataframe of the dataset). The input is a row from the dataset and a list of optional input parameters which will depend on your approach; `k` is the number of returned recommendations seems useful, though.

The output should be a `pd.DataFrame` containing the recommendations. The output dataframe should have the same columns as the row + any additional columns you deem important (e.g., any score or tags that you might want to add to your recommendations).

In principle, the method `get_top_recommendations()` may be imported from a external Python (.py) script as well.

In [21]:
def get_top_recommendations(row, **kwargs) -> pd.DataFrame:
    
    #####################################################
    ## Initialize the required parameters
    
    # The number of recommendations seem recommended
    # Additional input parameters are up to you
    topk = kwargs.get('topk')
    
    #for key, value in kwargs.items():
        #if key == 'topk':
          #topk = value
            
       
    #####################################################
    ## Compute your recommendations
    #

    df_copy = df_sample.copy()
    n = row.name
    one_sim = house_similarities[n]
    one_sim = np.delete(one_sim,obj = n)
    topk_index = one_sim.argsort()[-topk:len(one_sim)]

    df_result = df_copy.iloc[topk_index]
        
    # Return the dataset with the k recommendations
    return df_result


## Testing the Recommendation Engine

This will be the main part of your notebook to allow for testing your solutions. Most basically, for a given listing (defined by the row id in your input dataframe), we would like to see the recommendations you make. So however you set up your notebook, it should have at least a comparable section that will allow us to run your solution for different inputs.

### Pick a Sample Listing as Input

In [24]:
# Pick a row id of choice
row_id = 823
#row_id = 20
#row_id = 30
#row_id = 40
#row_id = 50

# Get the row from the dataframe (an valid row ids will throw an error)
row = df_sample.iloc[row_id]

# Just for printing it nicely, we create a new dataframe from this single row
pd.set_option('display.max_columns', None)
pd.DataFrame([row])

listing_id                                       title  \
823      152807  hdb flat for sale in 273b bishan street 24   

                                 address property_name property_type  \
823  ang mo kio / bishan / thomson (d20)   natura loft   hdb 5 rooms   

                tenure  built_year  num_beds  num_baths  size_sqft  \
823  99-year leasehold      2011.0         3          2       1292   

      furnishing available_unit_types  total_num_units  \
823  unspecified              3, 4 br       410.657534   

                                  property_details_url       lat        lng  \
823  https://www.99.co/singapore/hdb/natura-loft-de...  1.356085  103.84347   

     elevation    subzone planning_area      price  bed_bath_ratio  \
823          0  marymount        bishan  1249500.0             1.5   

     count_commerical  count_mrt_station  count_primary  count_secondary  \
823                 1                  3              0                2   

     count_shopping  
823               2

## Compute and Display the recommendations

Since the method `get_top_recommendations()` returns a `pd.DataFrame`, it's easy to display the result.

In [23]:
pd.set_option('display.max_columns', None)

k = 5

df_recommendations = get_top_recommendations(row, topk=k)

df_recommendations.head(k)

listing_id                                             title  \
11876      924141        3 bed condo for sale in kingsford waterbay   
3655       597014        5 bed condo for sale in cliveden at grange   
14993      583325  3 bed condo for sale in parc komo / komo shoppes   
13353      684205               6 bed house for sale in the carrara   
12656      467717        hdb flat for sale in 417 clementi avenue 1   

                                        address             property_name  \
11876                   74 upper serangoon view        kingsford waterbay   
3655                            106 grange road        cliveden at grange   
14993               965 upper changi road north  parc komo / komo shoppes   
13353                          mountbatten road               the carrara   
12656  buona vista / west coast / clementi (d5)             casa clementi   

             property_type             tenure  built_year  num_beds  \
11876                condo  99-year leasehold      2018.0         3   
3655                 condo           freehold      2011.0         5   
14993            apartment           freehold      2023.0         3   
13353  semi-detached house           freehold      2025.0         6   
12656          hdb 3 rooms  99-year leasehold      2013.0         3   

       num_baths  size_sqft   furnishing   available_unit_types  \
11876          2        893  unspecified       1, 2, 3, 4, 5 br   
3655           6       6000  unspecified          3, 4, 5, 6 br   
14993          3        980      partial       1, 2, 3, 4, 5 br   
13353          6       6000  unspecified  studio, 4, 5, 6, 7 br   
12656          2       1065  unspecified          1, 2, 3, 5 br   

       total_num_units                               property_details_url  \
11876       1165.00000  https://www.99.co/singapore/condos-apartments/...   
3655         110.00000  https://www.99.co/singapore/condos-apartments/...   
14993        276.00000  https://www.99.co/singapore/condos-apartments/...   
13353          6.00000  https://www.99.co/singapore/houses/the-carrara...   
12656        842.75841  https://www.99.co/singapore/hdb/casa-clementi-...   

            lat         lng  elevation         subzone  planning_area  \
11876  1.374566  103.903395          0         kangkar        hougang   
3655   1.299832  103.826298          0   one tree hill   river valley   
14993  1.357874  103.969563          0     changi west         changi   
13353  1.300620  103.892783          0     mountbatten  marine parade   
12656  1.310279  103.770762          0  clementi woods       clementi   

           price  bed_bath_ratio  count_commerical  count_mrt_station  \
11876  1543500.0        1.500000                 1                  0   
3655   9834300.0        0.833333                 0                  1   
14993  1912000.0        1.000000                 0                  0   
13353  9834300.0        1.000000                 0                  1   
12656  1016400.0        1.500000                 0                  2   

       count_primary  count_secondary  count_shopping  
11876              1                3               0  
3655               1                0              11  
14993              0                1               0  
13353              1                4               0  
12656              2                4               4

---
# Track2
### User Based CF

Randomly produce user profile.
With both positive ratings and negative ratings.

In [25]:
df_track2 = df_sample.copy()
df_track2['per_price'] = df_track2['price']/df_track2['size_sqft']
np.percentile(df_track2['per_price'],[25,50,75])

array([ 884.89208633, 1679.98234057, 2506.42300059])

Assume that user prefer `built_year` after 2000 and `per_price` lower than 2500(As 2506 is the 75% quantile of `per_price`).

In [26]:
np.random.seed(123)
#randomly generate 500 users' data

n = 500
#positive rating for select houses, for houses with per_price<=2500 & built_year>=2000.
user_select = df_track2[(df_track2.per_price<=2500)&(df_track2.built_year>=2000)]
user_item = user_select.sample(n=n)
pos_score = np.arange(3.0, 5.5, 0.5)
user_item_index = user_item['listing_id'].tolist()
user_item['rating'] = np.random.choice(pos_score,n)
user_item['user_id'] = range(n)

Assume that user give low ratings for houses `built_year` after 2000 and no mrt station near within 1 km.

In [27]:
#negative rating for select houses, for houses with count_mrt_station==0 & built_year<2000
user_select_neg = df_track2[(df_track2.count_mrt_station==0)&(df_track2.built_year<2000)]
user_item_neg = user_select_neg.sample(n=n)
neg_score = np.arange(0, 3.0, 0.5)
user_item_neg_index = user_item_neg['listing_id'].tolist()
user_item_neg['rating'] = np.random.choice(neg_score,n)
user_item_neg['user_id'] = range(n)

In [28]:
#combine the two data frame with user_id and ratings.
combine_data = pd.concat([user_item,user_item_neg])
combine_data.sample(5)

listing_id                                       title  \
14609      589066       4 bed house for sale in burgundy hill   
14335      630385    hdb flat for sale in 185 boon lay avenue   
3558       273160        6 bed house for sale in opera estate   
12211      455163  2 bed condo for sale in holland residences   
13377      987271       4 bed condo for sale in parc clematis   

                              address       property_name  \
14609               burgundy crescent       burgundy hill   
14335  boon lay / jurong / tuas (d22)      boon lay vista   
3558                   fidelio street        opera estate   
12211                   5 taman warna  holland residences   
13377                8f jalan lempeng       parc clematis   

             property_type             tenure  built_year  num_beds  \
14609  semi-detached house  99-year leasehold      1999.0         4   
14335          hdb 3 rooms  99-year leasehold      1975.0         2   
3558              bungalow           freehold      1980.0         6   
12211                condo           freehold      2012.0         2   
13377                condo  99-year leasehold      2023.0         4   

       num_baths  size_sqft   furnishing               available_unit_types  \
14609          5       3200  unspecified                   3, 4, 5, 6, 7 br   
14335          2        635  unspecified                      1, 2, 3, 5 br   
3558           7       6000  unspecified  studio, 2, 3, 4, 5, 6, 7, 8, 9 br   
12211          2        958  unspecified                         1, 2, 3 br   
13377          4       1496  unspecified           studio, 1, 2, 3, 4, 5 br   

       total_num_units                               property_details_url  \
14609       166.000000  https://www.99.co/singapore/houses/burgundy-hi...   
14335       580.477064  https://www.99.co/singapore/hdb/boon-lay-vista...   
3558        281.600000  https://www.99.co/singapore/condos-apartments/...   
12211        83.000000  https://www.99.co/singapore/condos-apartments/...   
13377      1468.000000  https://www.99.co/singapore/condos-apartments/...   

            lat         lng  elevation            subzone planning_area  \
14609  1.339176  103.760448          0  bukit batok south   bukit batok   
14335  1.346580  103.711182          0     boon lay place   jurong west   
3558   1.319664  103.923766          0            frankel         bedok   
12211  1.311628  103.793885          0      holland drive    queenstown   
13377  1.317138  103.759701          0              faber      clementi   

           price  bed_bath_ratio  count_commerical  count_mrt_station  \
14609  3024000.0        0.800000                 0                  0   
14335   459900.0        1.000000                 0                  0   
3558   8998500.0        0.857143                 0                  0   
12211  2320500.0        1.000000                 1                  3   
13377  2822400.0        1.000000                 0                  1   

       count_primary  count_secondary  count_shopping    per_price  rating  \
14609              2                0               0   945.000000     0.0   
14335              3                1               1   724.251969     2.0   
3558               2                1               2  1499.750000     0.0   
12211              0                0               3  2422.233820     3.0   
13377              3                2               3  1886.631016     4.5   

       user_id  
14609       44  
14335      300  
3558       299  
12211      388  
13377      152

In [29]:
num_house = df_track2.shape[0]
user_profiles = np.zeros((n, num_house))
M = np.zeros((n, num_house))
M.shape

(500, 18328)

In [30]:
#fill in with created ratings 
for index, row in combine_data.iterrows():
    user_id, house_id, rating = row['user_id'], row.name, row['rating']
    M[user_id][house_id] = rating

M[152][13377]

4.5

In [31]:
masked = np.ma.masked_equal(M, 0)
house_mean_ratings = np.mean(masked, axis=0)
user_mean_ratings = np.mean(masked, axis=1).reshape(-1, 1)
M_normalized = M - user_mean_ratings
M_normalized[M == 0] = 0
user_similarities = cosine_similarity(M_normalized, M_normalized)
user_similarities.shape

(500, 500)

## Computing the Top Recommendations

In [43]:
def get_top_recommendations(row, **kwargs) -> pd.DataFrame:
    
    #####################################################
    ## Initialize the required parameters
    
    topk = kwargs.get('topk')
    factor = kwargs.get('factor')  

            
       
    #####################################################
    ## Compute your recommendations
    #

    user_idx = range(n)

    recommendations = np.argsort(user_similarities[user_idx].squeeze())
    recommendations = recommendations[::-1][:topk*factor]
    recommendations = recommendations[np.random.choice(range(recommendations.shape[0]), size=topk,replace=False)]
    recom_index = recommendations[-1][1:topk+1]
    
    df_result = df_sample.iloc[recom_index]
        
    # Return the dataset with the k recommendations
    return df_result


## Testing the Recommendation Engine

In [45]:
# Pick a row id of choice
#row_id = 10
#row_id = 20
#row_id = 30
#row_id = 40
row_id = 50

# Get the row from the dataframe (an valid row ids will throw an error)
row = df_sample.iloc[row_id]

# Just for printing it nicely, we create a new dataframe from this single row
pd.DataFrame([row])

listing_id                     title                 address  \
50      720007  3 bed ec for sale in ola  70 anchorvale crescent   

   property_name    property_type             tenure  built_year  num_beds  \
50           ola  executive condo  99-year leasehold      2023.0         3   

    num_baths  size_sqft furnishing available_unit_types  total_num_units  \
50          2        926      fully        2, 3, 4, 5 br            548.0   

                                 property_details_url       lat         lng  \
50  https://www.99.co/singapore/condos-apartments/ola  1.397349  103.892353   

    elevation     subzone planning_area      price  bed_bath_ratio  \
50          0  anchorvale      sengkang  1076200.0             1.5   

    count_commerical  count_mrt_station  count_primary  count_secondary  \
50                 0                  1              5                2   

    count_shopping  
50               1

## Compute and Display the recommendations

In [46]:
k = 5

df_recommendations = get_top_recommendations(row, topk=k, factor=2)

df_recommendations.head(k)

listing_id                                              title  \
340      194614                3 bed condo for sale in 35 gilstead   
339      321215          hdb flat for sale in 120b edgedale plains   
338      362515  hdb flat for sale in 194a bukit batok west ave...   
337      931334           hdb flat for sale in 154 mei ling street   
336      928141            hdb flat for sale in 208 boon lay place   

                                               address  \
340                                   35 gilstead road   
339                 hougang / punggol / sengkang (d19)   
338  bukit batok / bukit panjang / choa chu kang (d23)   
337                      alexandra / commonwealth (d3)   
336                     boon lay / jurong / tuas (d22)   

               property_name property_type             tenure  built_year  \
340              35 gilstead         condo           freehold      2022.0   
339             punggol edge   hdb 3 rooms  99-year leasehold      2017.0   
338  skyline i @ bukit batok   hdb 4 rooms  99-year leasehold      2017.0   
337         mei ling heights   hdb 3 rooms  99-year leasehold      1970.0   
336           boon lay place   hdb 3 rooms  99-year leasehold      1976.0   

     num_beds  num_baths  size_sqft   furnishing available_unit_types  \
340         3          4       1539      partial        1, 2, 3, 4 br   
339         3          2       1044  unfurnished           2, 3, 4 br   
338         3          2       1022      partial        2, 3, 4, 5 br   
337         2          2        721        fully              2, 3 br   
336         2          2        700      partial      studio, 2, 3 br   

     total_num_units                               property_details_url  \
340        70.000000  https://www.99.co/singapore/condos-apartments/...   
339       688.000000  https://www.99.co/singapore/hdb/punggol-edge-d...   
338       426.836858  https://www.99.co/singapore/hdb/skyline-i-buki...   
337       791.798354  https://www.99.co/singapore/hdb/mei-ling-heigh...   
336       580.477064  https://www.99.co/singapore/hdb/boon-lay-place...   

          lat         lng  elevation           subzone planning_area  \
340  1.317359  103.838029          0          moulmein        novena   
339  1.394980  103.910789          0     punggol field       punggol   
338  1.344810  103.745662          0  bukit batok west   bukit batok   
337  1.293818  103.803665          0          mei chin    queenstown   
336  1.347253  103.713914          0    boon lay place   jurong west   

         price  bed_bath_ratio  count_commerical  count_mrt_station  \
340  3609900.0            0.75                 1                  3   
339   651000.0            1.50                 0                  0   
338   681200.0            1.50                 0                  1   
337   459900.0            1.00                 1                  1   
336   459900.0            1.00                 0                  1   

     count_primary  count_secondary  count_shopping  
340              2                1               3  
339              6                3               3  
338              3                2               1  
337              2                1               3  
336              4                3               1

---
# Track3 turicreate

### Item Based CF with `turicreate` Package


In [ ]:
!pip install turicreate

In [36]:
import turicreate as tc

Also use the combined data with user_id and ratings in Track2.

In [37]:
recom_data = tc.SFrame(combine_data)
recom_data.head()

Columns:
	listing_id	int
	title	str
	address	str
	property_name	str
	property_type	str
	tenure	str
	built_year	float
	num_beds	int
	num_baths	int
	size_sqft	int
	furnishing	str
	available_unit_types	str
	total_num_units	float
	property_details_url	str
	lat	float
	lng	float
	elevation	int
	subzone	str
	planning_area	str
	price	float
	bed_bath_ratio	float
	count_commerical	int
	count_mrt_station	int
	count_primary	int
	count_secondary	int
	count_shopping	int
	per_price	float
	rating	float
	user_id	int

Rows: 10

Data:
+------------+-------------------------------+-----------------------+
| listing_id |             title             |        address        |
+------------+-------------------------------+-----------------------+
|   457122   | 2 bed condo for sale in th... |     23 amber road     |
|   789538   | 1 bed condo for sale in fo... |    32 toh tuck road   |
|   151116   | 6 bed house for sale in pa... |  lorong 1 realty park |
|   554015   | 3 bed condo for sale in th... |    33 how sun road    |
|   821040   | 5 bed condo for sale in on... |    1 bernam street    |
|   881064   | 2 bed condo for sale in mi... |    8 hillview rise    |
|   662120   | 3 bed ec for sale in north... |    39 yishun close    |
|   199434   | 4 bed condo for sale in pi... | 3 northumberland road |
|   145928   | 3 bed condo for sale in ve... |    3 jalan kemaman    |
|   522539   | 7 bed house for sale in se... |    treasure island    |
+------------+-------------------------------+-----------------------+
+-------------------------------+-----------------+-------------------+------------+
|         property_name         |  property_type  |       tenure      | built_year |
+-------------------------------+-----------------+-------------------+------------+
|       the aristo @ amber      |    apartment    |      freehold     |   2013.0   |
|      forett @ bukit timah     |      condo      |      freehold     |   2024.0   |
|      parkwood collection      |  cluster house  | 99-year leasehold |   2022.0   |
|           the lilium          |      condo      |      freehold     |   2021.0   |
|           one bernam          |      condo      | 99-year leasehold |   2026.0   |
|            midwood            |      condo      | 99-year leasehold |   2024.0   |
|           north gaia          | executive condo | 99-year leasehold |   2027.0   |
| piccadilly grand / piccadi... |      condo      | 99-year leasehold |   2026.0   |
|            verticus           |      condo      |      freehold     |   2024.0   |
|          sentosa cove         |     bungalow    | 99-year leasehold |   2010.0   |
+-------------------------------+-----------------+-------------------+------------+
+----------+-----------+-----------+-------------+----------------------------+
| num_beds | num_baths | size_sqft |  furnishing |    available_unit_types    |
+----------+-----------+-----------+-------------+----------------------------+
|    2     |     1     |    646    | unspecified |            2 br            |
|    1     |     1     |    517    | unspecified |  studio, 1, 2, 3, 4, 5 br  |
|    6     |     6     |    4865   | unspecified |         4, 5, 6 br         |
|    3     |     3     |    1227   | unfurnished |   studio, 2, 3, 4, 5 br    |
|    5     |     6     |    4306   | unspecified |  studio, 1, 2, 3, 4, 5 br  |
|    2     |     2     |    700    | unspecified |       1, 2, 3, 4 br        |
|    3     |     2     |    1076   | unfurnished |       2, 3, 4, 5 br        |
|    4     |     3     |    1389   | unspecified |  studio, 1, 2, 3, 4, 5 br  |
|    3     |     3     |    1152   | unspecified |       1, 2, 3, 4 br        |
|    7     |     10    |    6000   | unspecified | 2, 3, 4, 5, 6, 7, 8, 10 br |
+----------+-----------+-----------+-------------+----------------------------+
+--------------------+-------------------------------+--------------------+
|  total_num_units   |      property_details_url     |        lat         |
+--------------------+

## Computing the Top Recommendations

In [38]:
def get_top_recommendations(row, **kwargs) -> pd.DataFrame:
    
    #####################################################
    ## Initialize the required parameters
    
    topk = kwargs.get('topk')
    df = kwargs.get('df')
    users = row.user_id
            
       
    #####################################################
    ## Compute your recommendations
    #

    df_copy = df.copy()
    recom_data = tc.SFrame(df_copy)
    model = tc.item_similarity_recommender.create(recom_data, user_id='user_id', item_id='listing_id', target='rating', similarity_type='cosine', verbose=False)
    listing_id = model.recommend(users=[users],k=topk)['listing_id']

    df_result = combine_data[combine_data['listing_id'].apply(lambda i: i in listing_id)]
        
    # Return the dataset with the k recommendations
    return df_result

## Testing the Recommendation Engine

In [47]:
# Pick a row id of choice
row_id = 123
#row_id = 20
#row_id = 30
#row_id = 40
#row_id = 50

# Get the row from the dataframe (an valid row ids will throw an error)
row = combine_data.iloc[row_id]

# Just for printing it nicely, we create a new dataframe from this single row
pd.DataFrame([row])

listing_id                                title          address  \
3440      727180  1 bed condo for sale in the gazania  5 how sun drive   

     property_name property_type    tenure  built_year  num_beds  num_baths  \
3440   the gazania         condo  freehold      2022.0         1          1   

      size_sqft   furnishing      available_unit_types  total_num_units  \
3440        517  unspecified  studio, 1, 2, 3, 4, 5 br            250.0   

                                   property_details_url       lat        lng  \
3440  https://www.99.co/singapore/condos-apartments/...  1.344334  103.87869   

      elevation           subzone planning_area      price  bed_bath_ratio  \
3440          0  upper paya lebar     serangoon  1155000.0             1.0   

      count_commerical  count_mrt_station  count_primary  count_secondary  \
3440                 0                  3              3                5   

      count_shopping    per_price  rating  user_id  
3440               1  2234.042553     3.0      123

## Compute and Display the recommendations

In [42]:
get_top_recommendations(row,topk = 5,df = combine_data)

listing_id                                         title  \
11977      457122    2 bed condo for sale in the aristo @ amber   
8966       789538  1 bed condo for sale in forett @ bukit timah   
5404       151116   6 bed house for sale in parkwood collection   
1373       554015            3 bed condo for sale in the lilium   
11053      821040            5 bed condo for sale in one bernam   

                    address         property_name  property_type  \
11977         23 amber road    the aristo @ amber      apartment   
8966       32 toh tuck road  forett @ bukit timah          condo   
5404   lorong 1 realty park   parkwood collection  cluster house   
1373        33 how sun road            the lilium          condo   
11053       1 bernam street            one bernam          condo   

                  tenure  built_year  num_beds  num_baths  size_sqft  \
11977           freehold      2013.0         2          1        646   
8966            freehold      2024.0         1          1        517   
5404   99-year leasehold      2022.0         6          6       4865   
1373            freehold      2021.0         3          3       1227   
11053  99-year leasehold      2026.0         5          6       4306   

        furnishing      available_unit_types  total_num_units  \
11977  unspecified                      2 br             56.0   
8966   unspecified  studio, 1, 2, 3, 4, 5 br            633.0   
5404   unspecified                4, 5, 6 br             53.0   
1373   unfurnished     studio, 2, 3, 4, 5 br             80.0   
11053  unspecified  studio, 1, 2, 3, 4, 5 br            351.0   

                                    property_details_url       lat  \
11977  https://www.99.co/singapore/condos-apartments/...  1.300183   
8966   https://www.99.co/singapore/condos-apartments/...  1.339338   
5404   https://www.99.co/singapore/houses/parkwood-co...  1.367030   
1373   https://www.99.co/singapore/condos-apartments/...  1.345552   
11053  https://www.99.co/singapore/condos-apartments/...  1.273959   

              lng  elevation            subzone  planning_area      price  \
11977  103.900824          0      marine parade  marine parade  1312500.0   
8966   103.763893          0  bukit batok south    bukit batok  1050000.0   
5404   103.889966          0    hougang central        hougang  4014200.0   
1373   103.881514          0   upper paya lebar      serangoon  2705600.0   
11053  103.843635          0              anson  downtown core  9834300.0   

       bed_bath_ratio  count_commerical  count_mrt_station  count_primary  \
11977        2.000000                 0                  0              1   
8966         1.000000                 0                  0              1   
5404         1.000000                 0                  2              2   
1373         1.000000                 0                  1              3   
11053        0.833333                 0                  3              1   

       count_secondary  count_shopping    per_price  rating  user_id  
11977                1               5  2031.733746     3.5        0  
8966                 0               0  2030.947776     3.5        1  
5404                 3               3   825.118191     4.5        2  
1373                 5               1  2205.052975     4.0        3  
11053                0               1  2283.859731     4.5        4

---